tianshou/policy/modelbase/psrl.py

```diff
@@ -0,0 +1,211 @@
+
+from tianshou.policy import BasePolicy
+from tianshou.data import Batch, ReplayBuffer
+
+
+class PSRLModel(object):
+    """Implementation of Posterior Sampling Reinforcement Learning Model.
+
+    :param np.ndarray p_prior: dirichlet prior (alphas), with shape
+        (n_action, n_state, n_state).
+    :param np.ndarray rew_mean_prior: means of the normal priors of rewards,
+        with shape (n_state, n_action).
+    :param np.ndarray rew_std_prior: standard deviations of the normal priors
+        of rewards, with shape (n_state, n_action).
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.BasePolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        trans_count_prior: np.ndarray,
+        rew_mean_prior: np.ndarray,
+        rew_std_prior: np.ndarray,
+    ) -> None:
+        self.__eps = np.finfo(np.float32).eps.item()
+        self.trans_count = trans_count_prior
+        self.trans_count_prior_sum = np.sum(trans_count_prior, axis=2)
+        self.n_action, self.n_state, _ = trans_count_prior.shape
+        self.rew_mean = rew_mean_prior
+        self.rew_std = rew_std_prior
+        self.rew_count = np.zeros_like(rew_mean_prior) + self.__eps
+        self.policy: Optional[np.ndarray] = None
+        self.updated = False
+
+    def observe(
+        self, trans_count: np.ndarray,
+            rew_sum: np.ndarray, rew_count: np.ndarray
+    ) -> None:
+        """Add data into memory pool.
+
+        For rewards, we have a normal prior at first. After we observed a
+        reward for a given state-action pair, we use the mean value of our
+        observations instead of the prior mean as the posterior mean. The
+        standard deviations are in inverse proportion to the number of the
+        corresponding observations.
+
+        :param np.ndarray trans_count: the number of observations, with shape
+            (n_action, n_state, n_state).
+        :param np.ndarray rew_sum: total rewards, with shape
+            (n_state, n_action).
+        :param np.ndarray rew_count: the number of rewards, with shape
+            (n_state, n_action).
+        """
+        self.updated = False
+        self.trans_count += trans_count
+        sum_count = self.rew_count + rew_count
+        self.rew_mean = (self.rew_mean * self.rew_count + rew_sum) / sum_count
+        self.rew_std *= self.rew_count / sum_count
+        self.rew_count = sum_count
+        if np.sum(self.trans_count) > \
+                np.sum(self.trans_count_prior_sum) + 1000:
+            min_index = np.argmin(np.sum(self.trans_count, axis=2), axis=1)
+            mask = np.isclose(np.sum(self.trans_count, axis=2),
+                              self.trans_count_prior_sum).astype("float32")
+            self.trans_count[np.array(range(self.n_action)),
+                             min_index, min_index] += \
+                mask[np.array(range(self.n_action)), min_index]
```

Comment:
What are these lines doing? Could you please add some comment `# xxx` in L65?
And why there is a magic number 1000?